File: src/main/scala/hr/caellian/math/vector/VectorD.scala
/*
 * The MIT License (MIT)
 * Copyright (c) 2016 Tin Švagelj <tin.svagelj.email@gmail.com> a.k.a. Caellian
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package hr.caellian.math.vector

import java.nio.{Buffer, ByteBuffer, ByteOrder, DoubleBuffer}

import hr.caellian.math.matrix.{Matrix, MatrixD}
import hr.caellian.math.util.DataUtil

/**
  * @author caellian
  *         Created on 31/07/16 at 17:05 CET.
  */
/**
  * Vector class for double N-dimensional vectors.
  *
  * @author Caellian
  */
class VectorD(override val data: Array[Double]) extends Vector[Double] {
  /**
    * @return new conjugated vector.
    */
  def unary_- = {
    new VectorD(asArray.map(-_))
  }

  /**
    * Adds two vectors together and returns resulting vector.
    * In order to add to matrices together, they must be of same size.
    *
    * @param other vector to add to this one.
    * @return result of vector addition.
    */
  def +(other: Vector[Double]): VectorD = {
    require(size == other.size, "Invalid argument vector size!")
    new VectorD(data zip other.data map Function.tupled(_ + _))
  }

  /**
    * Subtracts other vector from this one and returns resulting vector.
    * In order to subtract one vector from another, both vectors must be of same size.
    *
    * @param other vector to subtract from this one.
    * @return result of vector subtraction.
    */
  def -(other: Vector[Double]): VectorD = {
    require(size == other.size, "Invalid argument vector size!")
    new VectorD(data zip other.data map Function.tupled(_ - _))
  }

  /**
    * Multiplies two vectors together and returns resulting vector.
    * In order to add to multiply vectors together, they must be of same size.
    *
    * @param other vector to multiply with this one.
    * @return result of vector multiplication.
    */
  def *(other: Vector[Double]): VectorD = {
    require(size == other.size, "Invalid argument vector size!")
    new VectorD(data zip other.data map Function.tupled(_ * _))
  }

  /**
    * Divides this vector with other and returns resulting vector.
    * In order to divide one vector from another, both vectors must be of same size.
    *
    * @param other vector dividend.
    * @return result of vector division.
    */
  def /(other: Vector[Double]): VectorD = {
    require(size == other.size, "Invalid argument vector size!")
    new VectorD(data zip other.data map Function.tupled(_ / _))
  }

  /**
    * Performs scalar addition on this vector and returns resulting vector.
    *
    * @param value scalar value.
    * @return result of scalar vector addition.
    */
  def +(value: Double): VectorD = {
    new VectorD(asArray.map(_ + value))
  }

  /**
    * Performs scalar subtraction on this vector and returns resulting vector.
    *
    * @param value scalar value.
    * @return result of scalar vector subtraction.
    */
  def -(value: Double): VectorD = {
    new VectorD(asArray.map(_ - value))
  }

  /**
    * Performs scalar multiplication on this vector and returns resulting vector.
    *
    * @param value scalar value.
    * @return result of scalar vector multiplication.
    */
  def *(value: Double): VectorD = {
    new VectorD(asArray.map(_ * value))
  }

  /**
    * Performs scalar division on this vector and returns resulting vector.
    *
    * @param value scalar value.
    * @return result of scalar vector division.
    */
  def /(value: Double): VectorD = {
    new VectorD(asArray.map(_ / value))
  }

  /**
    * @return max value of member of this vector.
    */
  def max: Double = {
    data.toList.max
  }

  /**
    * @return new vector containing absolute values of this vector.
    */
  def absolute: VectorD = {
    new VectorD(asArray.map(Math.abs))
  }

  /**
    * @return new vector with normalized values of this one.
    */
  def normalized: VectorD = {
    this / this.magnitude
  }

  /**
    * @return magnitude of this vector.
    */
  def magnitude: Double = {
    distanceTo(this)
  }

  /**
    * Calculates distance from this to other vector.
    *
    * @param other vector to calculate distance to.
    * @return distance between this and other vector.
    */
  def distanceTo(other: Vector[Double]): Double = {
    Math.sqrt(this dot other)
  }

  /**
    *
    * @param other other vector used to determine dot product.
    * @return dot product of two vectors
    */
  def dot(other: Vector[Double]): Double = {
    require(size == other.size, "Vectors must be of same size.")
    (data zip other.data map Function.tupled(_ * _)).sum
  }

  /**
    * Returns cross product of this and other vector.
    *
    * @param other other vector used to determine cross product.
    * @return cross product.
    */
  def cross(other: Vector[Double]): VectorD = {
    require(size == other.size, "Vectors must be of same size.")
    // Some cross product analogs do exist for 2D space, but they are inconsistent.
    require(size == 7 || size == 3, s"Cross product does not exist in $size-dimensional space!")

    val result = Array.ofDim[Double](this.size)
    this.size match {
      case 7 =>
        result(X) = this (Y) * other(W) - this (W) * other(Y) + this (Z) * other(6) - this (6) * other(Z) + this (4) * other(5) - this (5) * other(4)
        result(Y) = this (Z) * other(4) - this (4) * other(Z) + this (W) * other(X) - this (X) * other(W) + this (5) * other(6) - this (6) * other(5)
        result(Z) = this (W) * other(5) - this (5) * other(W) + this (4) * other(Y) - this (Y) * other(4) + this (6) * other(X) - this (X) * other(6)
        result(W) = this (4) * other(6) - this (6) * other(4) + this (5) * other(Z) - this (Z) * other(5) + this (X) * other(Y) - this (Y) * other(X)
        result(4) = this (5) * other(X) - this (X) * other(5) + this (6) * other(W) - this (W) * other(6) + this (Y) * other(Z) - this (Z) * other(Y)
        result(5) = this (6) * other(Y) - this (Y) * other(6) + this (X) * other(4) - this (4) * other(X) + this (Z) * other(W) - this (W) * other(Z)
        result(6) = this (X) * other(Z) - this (Z) * other(X) + this (Y) * other(5) - this (5) * other(Y) + this (W) * other(4) - this (4) * other(W)
      case 3 =>
        result(X) = this (Y) * other(Z) - this (Z) * other(Y)
        result(Y) = this (Z) * other(X) - this (X) * other(Z)
        result(Z) = this (X) * other(Y) - this (Y) * other(X)
    }
    new VectorD(result)
  }

  /**
    * Rotates this vector using rotation matrix.
    *
    * @param rotationMatrix rotation matrix to use.
    * @return rotated vector.
    */
  def rotated(rotationMatrix: Matrix[Double]): VectorD = {
    (rotationMatrix * verticalMatrix).toVector.asInstanceOf[VectorD]
  }

  /**
    * Linearly interpolates between two vectors.
    *
    * @param destination vector with destination coordinates.
    * @param percent     percentage of
    * @return linear interpolation.
    */
  def lerp(destination: Vector[Double], percent: Double): VectorD = {
    this + (destination - this) * percent
  }

  /**
    * Vertical matrix containing data of this vector.
    */
  val verticalMatrix = MatrixD(true, asArray)

  /**
    * Horizontal matrix containing data of this vector.
    */
  val horizontalMatrix = MatrixD(false, asArray)

  /**
    * @return buffer containing data of this vector.
    */
  def asBuffer: Buffer = {
    ByteBuffer.allocateDirect(data.length << 2).order(ByteOrder.nativeOrder).asDoubleBuffer().put(data).flip
  }

  /**
    * @return clone of this vector.
    */
  override def replicated: VectorD = new VectorD(asArray)

  /**
    * Creates a new instance of wrapper containing given data.
    *
    * @param data data of new wrapper.
    * @return new instance of wrapper containing argument data.
    */
  override def withData(data: Array[Double]): VectorD = new VectorD(data)

  /**
    * Vector hashcode depends on vector data and will change is vector data is modified!
    *
    * @return hashcode of this vector.
    */
  override def hashCode: Int = {
    DataUtil.hashCode(data)
  }
}

/**
  * Object used to create {@link VectorD} vectors and ensure their consistency.
  * Use initializer functions from this object only if there is no alternative ones in more specific objects as some
  * of these are very CPU intensive.
  *
  * @author Caellian
  */
object VectorD {
  type VecD = VectorD

  /**
    * Creates a new vector using given values.
    *
    * @param values values to create a new vector from.
    * @return created vector.
    */
  def apply(values: Double*): VectorD = {
    new VectorD(values.toArray[Double])
  }

  /**
    * Creates a new vector using array values.
    *
    * @param values value array to create a new vector from.
    * @return created vector.
    */
  def apply(values: Array[Double]): VectorD = {
    new VectorD(values)
  }

  /**
    * Creates a new vector using buffer values.
    *
    * @param buffer buffer to create a new vector from.
    * @return created vector.
    */
  def apply(buffer: DoubleBuffer): VectorD = {
    new VectorD((0 until buffer.capacity()).map(buffer.get).toArray)
  }
}

File: src/main/kotlin/hr/caellian/math/vector/VectorD.kt
package hr.caellian.math.vector

import hr.caellian.math.matrix.Matrix
import hr.caellian.math.matrix.MatrixD
import java.nio.Buffer
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.DoubleBuffer
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Vector class for double N-dimensional vectors.
 *
 * @param wrapped value array to create a new vector from.
 *
 * @author Caellian
 */
class VectorD(override var wrapped: Array<Double> = emptyArray()) : VectorN<Double>() {

    /**
     * Creates a new vector using given values.
     *
     * @param values values to create a new vector from.
     * @return created vector.
     */
    constructor(vararg values: Double) : this(values.toTypedArray())

    /**
     * Creates a new vector using given collection values.
     *
     * @param values collection values to create a new vector from.
     * @return created vector.
     */
    constructor(values: Collection<Double>) : this(values.toTypedArray())

    /**
     * Creates a new vector using buffer values.
     *
     * @param buffer buffer to create a new vector from.
     * @return created vector.
     */
    constructor(buffer: DoubleBuffer) : this((0 until buffer.capacity()).map { buffer.get() })

    /**
     * @return new vector with negated values.
     */
    override operator fun unaryMinus(): VectorD = VectorD(Array(size) { -this[it] })

    /**
     * Adds two vectors together and returns resulting vector.
     * In order to add to matrices together, they must be of same size.
     *
     * @return result of vector addition.
     */
    override operator fun plus(other: Vector<Double>): VectorD {
        require(size == other.size) { "Invalid argument vector size: ${other.size}!" }
        return VectorD(Array(size) { this[it] + other[it] })
    }

    /**
     * Subtracts other vector from this one and returns resulting vector.
     * In order to subtract one vector from another, both vectors must be of same size.
     *
     * @return result of vector subtraction.
     */
    override operator fun minus(other: Vector<Double>): VectorD {
        require(size == other.size) { "Invalid argument vector size: ${other.size}!" }
        return VectorD(Array(size) { this[it] - other[it] })
    }

    /**
     * Multiplies two vectors together and returns resulting vector.
     * In order to add to multiply vectors together, they must be of same size.
     *
     * @return result of vector multiplication.
     */
    override operator fun times(other: Vector<Double>): VectorD {
        require(size == other.size) { "Invalid argument vector size: ${other.size}!" }
        return VectorD(Array(size) { this[it] * other[it] })
    }

    /**
     * Divides this vector with other and returns resulting vector.
     * In order to divide one vector with another, both vectors must be of same size.
     *
     * @return result of vector division.
     */
    override operator fun div(other: Vector<Double>): VectorD {
        require(size == other.size) { "Invalid argument vector size: ${other.size}!" }
        return VectorD(Array(size) { this[it] / other[it] })
    }

    /**
     * Performs scalar addition on this vector and returns resulting vector.
     *
     * @return result of vector scalar addition.
     */
    override operator fun plus(value: Double): VectorD = VectorD(Array(size) { this[it] + value })

    /**
     * Performs scalar subtraction on this vector and returns resulting vector.
     *
     * @return result of scalar vector subtraction.
     */
    override operator fun minus(value: Double): VectorD = VectorD(Array(size) { this[it] - value })

    /**
     * Performs scalar multiplication on this vector and returns resulting vector.
     *
     * @return result of scalar vector multiplication.
     */
    override operator fun times(value: Double): VectorD = VectorD(Array(size) { this[it] * value })

    /**
     * Performs scalar division on this vector and returns resulting vector.
     *
     * @return result of scalar vector division.
     */
    override operator fun div(value: Double): VectorD = VectorD(Array(size) { this[it] / value })

    /**
     * @return biggest value of a member of this vector.
     */
    override fun max(): Double = wrapped.max() ?: 0.0

    /**
     * @return smalled value of a member of this vector.
     */
    override fun min(): Double = wrapped.min() ?: 0.0

    /**
     * @return new vector containing absolute values of this vector.
     */
    override fun absolute(): VectorD = VectorD(Array(size) { abs(this[it]) })

    /**
     * @return new vector with normalized values of this one.
     */
    override fun normalized(): VectorD = this / magnitude()

    /**
     * @return magnitude of this vector.
     */
    override fun magnitude(): Double = distanceTo(this)

    /**
     * Calculates distance from this to other vector.
     *
     * @return distance between this and other vector.
     */
    override fun distanceTo(other: Vector<Double>): Double = sqrt(this dot other)

    /**
     *
     * @return dot product of two vectors.
     */
    override fun dot(other: Vector<Double>): Double {
        require(size == other.size) { "Invalid argument vector size: ${other.size}!" }
        return wrapped.zip(other.wrapped).sumByDouble { (a, b) -> a * b }
    }

    /**
     * Returns cross product of this and other vector.
     *
     * @return cross product.
     */
    override fun cross(other: Vector<Double>): VectorD {
        require(size == other.size) { "Invalid argument vector size: ${other.size}!" }

        return when (size) {
            3 -> VectorD(arrayOf(
                    this[1] * other[2] - this[2] * other[1],
                    this[2] * other[0] - this[0] * other[2],
                    this[0] * other[1] - this[1] * other[0]
            ))
            7 -> VectorD(arrayOf(
                    this[1] * other[3] - this[3] * other[1] + this[2] * other[6] - this[6] * other[2] + this[4] * other[5] - this[5] * other[4],
                    this[2] * other[4] - this[4] * other[2] + this[3] * other[0] - this[0] * other[3] + this[5] * other[6] - this[6] * other[5],
                    this[3] * other[5] - this[5] * other[3] + this[4] * other[1] - this[1] * other[4] + this[6] * other[0] - this[0] * other[6],
                    this[4] * other[6] - this[6] * other[4] + this[5] * other[2] - this[2] * other[5] + this[0] * other[1] - this[1] * other[0],
                    this[5] * other[0] - this[0] * other[5] + this[6] * other[3] - this[3] * other[6] + this[1] * other[2] - this[2] * other[1],
                    this[6] * other[1] - this[1] * other[6] + this[0] * other[4] - this[4] * other[0] + this[2] * other[3] - this[3] * other[2],
                    this[0] * other[2] - this[2] * other[0] + this[1] * other[5] - this[5] * other[1] + this[3] * other[4] - this[4] * other[3]
            ))
            else -> throw NotImplementedError("Cross product does not exist in $size-dimensional space!")
        }
    }

    /**
     * Rotates this vector using rotation matrix.
     *
     * @return rotated vector.
     */
    override fun rotated(rotationMatrix: Matrix<Double>): VectorD = (rotationMatrix * verticalMatrix).toVector() as VectorD

    /**
     * Linearly interpolates between two vectors.
     *
     * @return linear interpolation.
     */
    override fun lerp(destination: Vector<Double>, percent: Double): VectorD = this + (destination - this) * percent

    /**
     * Vertical matrix containing data of this vector.
     */
    override val verticalMatrix: MatrixD by lazy { MatrixD(arrayOf(toArray()), true) }
    /**
     * Horizontal matrix containing data of this vector.
     */
    override val horizontalMatrix: MatrixD by lazy { MatrixD(arrayOf(toArray())) }

    /**
     * Returns array containing vector data.
     *
     * @return array containing data of this vector.
     */
    override fun toArray(): Array<Double> = Array(size) { wrapped[it] }

    /**
     * @return clone of this vector.
     */
    override fun replicated(): VectorD = VectorD(toArray())

    /**
     * Creates a new instance of wrapper containing given data.
     *
     * @param data data of new wrapper.
     * @return new instance of wrapper containing argument data.
     */
    override fun withData(wrapped: Array<Double>): VectorD = VectorD(wrapped)

    /**
     * @return [Buffer] containing data of represented object.
     */
    override fun toBuffer(): Buffer {
        val result = ByteBuffer.allocateDirect(wrapped.size shl 2).order(
                ByteOrder.nativeOrder()).asDoubleBuffer()
        wrapped.forEach { result.put(it) }
        return result.flip()
    }
}
